inbm/dispatcher-agent/dispatcher/sota/oras_util.py

```diff
@@ -0,0 +1,150 @@
+"""
+    ORAS tool will be called by dispatcher to perform the image downloading in TiberOS.
+
+    Copyright (C) 2017-2024 Intel Corporation
+    SPDX-License-Identifier: Apache-2.0
+"""
+import logging
+import os
+import requests
+import json
+from urllib.parse import urlsplit
+from typing import Optional
+from dispatcher.dispatcher_exception import DispatcherException
+from inbm_common_lib.shell_runner import PseudoShellRunner
+from inbm_common_lib.utility import CanonicalUri
+from dispatcher.packagemanager.local_repo import DirectoryRepo
+from dispatcher.packagemanager.package_manager import verify_source
+from ..packagemanager.irepo import IRepo
+from ..dispatcher_broker import DispatcherBroker
+from ..constants import CACHE
+
+
+logger = logging.getLogger(__name__)
+
+
+def oras_download(dispatcher_broker: DispatcherBroker, uri: CanonicalUri,
+             repo: IRepo, username: Optional[str], password: str, umask: int) -> None:
+    """Downloads files and places capsule file in path mentioned by manifest file.
+
+    @param dispatcher_broker: DispatcherBroker object used to communicate with other INBM services
+    @param uri: URI of the source location
+    @param repo: repository for holding the download
+    @param username: username to use for download
+    @param password: password to use for download
+    @param umask: file permission mask
+    @raises DispatcherException: any exception
+    """
+    dispatcher_broker.telemetry(f'Package to be fetched from {uri.value}')
+    dispatcher_broker.telemetry(
+        'Checking authenticity of package by checking signature and source')
+
+    if not isinstance(uri, CanonicalUri):
+        raise DispatcherException("Internal error: uri improperly passed to download function")
+
+    """
+    In case of uri.value = https://registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest
+    source = https://registry-rs.internal.ledgepark.intel.com/one-intel-edge
+    registry_server = registry-rs.internal.ledgepark.intel.com
+    image = tiberos
+    image_tag = latest
+    image_full_path = registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest
+    repository_name = one-intel-edge
+    registry_manifest = https://registry-rs.internal.ledgepark.intel.com/v2/one-intel-edge/tiberos/manifest/latest
+    """
+    try:
+        source = uri.value[:-(len(uri.value.split('/')[-1]) + 1)]
```

Comment:
use `urllib.parse.urlparse`, e.g.

```
parsed_uri = urlparse(uri.value)
registry_server = parsed_uri.netloc
path_parts = parsed_uri.path.strip('/').split('/')
if len(path_parts) < 2:
    raise DispatcherException(f"Invalid URI format: {uri.value}")
repository_name = '/'.join(path_parts[:-1])
image = path_parts[-1].split(':')[0]
image_tag = path_parts[-1].split(':')[1] if ':' in path_parts[-1] else 'latest'
```

Also: split out the parsing part of this function so it can be unit tested with various URIs.

Reply:
Added this method and added unit tests.